### ISSUE TITLE
Include Technical Overview Documentation using mkdocs

### ISSUE THREAD

**Issue by idotobi** [xournalpp/xournalpp#2538]
**Is your feature request related to a problem? Please describe.**
It is hard to get an overview of the project by only reading the code and the documentation that is already present.
Hence, it would be nice to have an easy to edit, well integrated way to document higher level concepts/structures/infrastructure used for the xournalpp project.

**Describe the solution you'd like**
Using mkdocs as part of the repository such that the mental burden of writing documentation while developing is minimal.

**Describe alternatives you've considered**
 - Using Github wiki:
    - separate repo
    - cannot be build locally
    - no offline search

**Additional context**

Pointers: https://www.mkdocs.org/user-guide/deploying-your-docs/#project-pages

**Open Questions**

mkdocs could either be deployed using [Github Project Pages](https://docs.github.com/en/free-pro-team@latest/github/working-with-github-pages/about-github-pages#project-pages-sites) or [Readthedocs](https://docs.readthedocs.io/en/stable/intro/getting-started-with-mkdocs.html). Which one is preferable is not yet clear.

**Comment by idotobi** [comment]
I just played around with Readthedocs and it's actually quite simple: 
https://xournalpp.readthedocs.io/en/latest/ (built from https://github.com/idotobi/xournalpp/tree/mkdocs-dev-documentation).

The only thing I don't quite like is that the docs may contain advertisement and that I didn't need to proof anything to claim xournalpp as my project :thinking: 
Apart from that it looks quite nice. What do you think?

**Comment by idotobi** [comment]
I guess another question would be how we structure the directory while having also the other `doc` folder.
We could go for something like this

```
 - rest of code
 - readthedocs.yml     # needs to be on project root
 - doc                          # other documentation
 - dev-docs
    - mkdocs.yml
    - docs
        - actual content
```

or 

```
 - rest of code
 - readthedocs.yml     # needs to be on project root
 - mkdocs.yml
 - doc                          # other documentation
 - dev-docs
      - actual content
```

**Comment by rolandlo** [comment]
It does look nice. If there is not more advertisement than the one seen, I don't mind about that. So what did you do to build the page? 
The second suggestion to structure the directory looks more clean to me.

**Comment by bhennion** [comment]
1. This is a great idea!
2. I think it would be better to have mkdocs directly on the github page (don't know how easy it is to do). It would probably get us rid of the add too.

**Comment by idotobi** [comment]
That's how github pages would look like https://idotobi.github.io/xournalpp/.

For the reproduction Steps:

**Readthedocs**:
* get account
* create project
* point project to repo
* configure (max. 5min)
* build

**Github pages**:
* install mkdocs (`pip install --user mkdocs`)
* run `mkdocs gh-deploy` (this pushes the built site to the branch `gh-pages` which is used by github pages)

I guess the Github pages option should be configured via some [Github actions](https://medium.com/dev-genius/continuous-deployment-for-mkdocs-with-github-actions-7dceec87f0ea) but apart from that it's also a nice approach. I don't have to strong an opinion.

**Comment by rolandlo** [comment]
The github pages are even better (no advertisements). You have created a user/organization github page, though, haven't you? There can only be one of those for each user and organization. We already have [https://xournalpp.github.io/xournalpp/](https://xournalpp.github.io/xournalpp/) as our website. A project page would be accessible under a name like [https://xournalpp.github.io/xournalpp/dev-docs/](https://xournalpp.github.io/xournalpp/dev-docs/).

**Comment by idotobi** [comment]
No, actually I created a project page, at least according to the mkdocs [docs](https://www.mkdocs.org/user-guide/deploying-your-docs/#project-pages).

So I guess this is what is also used for https://xournalpp.github.io/xournalpp/. It might make sense to use the user/organization github page for the xournalpp website, but I'm not sure if that would result in a different domain for the page. :thinking:

**Comment by rolandlo** [comment]
As I understand now, if we create a repository called `dev-docs` inside the [xournalpp organization](https://github.com/xournalpp), the project page created by `gh-deploy` will be available under [https://xournalpp.github.io/dev-docs/](https://xournalpp.github.io/dev-docs/), whereas the organization webpage will still be [https://xournalpp.github.io](https://xournalpp.github.io). This looks like the proper solution.

**Comment by rolandlo** [comment]
Automatically running gh-deploy could probably also configured via the same github actions as on the website: [https://github.com/xournalpp/xournalpp.github.io/blob/site/.github/workflows/deploy.yml](https://github.com/xournalpp/xournalpp.github.io/blob/site/.github/workflows/deploy.yml)

**Comment by Technius** [comment]
I could set this quickly up over the weekend, since I've already done it before.

**Comment by idotobi** [comment]
@Technius : Did you get to do this over Christmas?

If not I could also offer to try it out and you could review?  I think I would have some time over the weekend.

**Comment by idotobi** [comment]
@Technius , @rolandlo : If you don't mind I would take this up again.

I would try to actually put the documentation in this repo. At my current work we have the docs as part of the production repo, and it's working really nice. This way the code and the docs stay in sync as long as reviewers of PRs also keep in mind to ask for doc additions if it makes sense. Would that be OK with you?

Are there any requirements you would have on the doc infrastructure that I should keep in mind?

**Comment by Technius** [comment]
One concern I have is how we would deploy the docs (or if we would want to at all). We could deploy to gh-pages directly from this repo, but then the published URL would be `xournalpp.github.io/xournalpp`. Maybe it's easier to assume that developers interested in reading the docs would build the documentation site themselves.

I'm also worried about what CI will do if we push frequent doc changes; we don't want automated nightly builds and other redundant CI builds occurring because someone added a sentence of text to the documentation.

**Comment by idotobi** [comment]
@Technius :  I would assume that initially asking developers to build the docs locally is fine. Still I would also check out how we could maybe deploy it under a different url with gh-pages. Maybe there's a way.

With regards to the CI that's a valid concern. There must be an option to ignore changes in certain directories when triggering the CI. I'll try to figure this out as part of the issue then :nerd_face:
